server/coordinator/shard_picker.go

```diff
@@ -29,67 +27,68 @@ func NewRandomBalancedShardPicker() ShardPicker {
 }
 
 // PickShards will pick a specified number of shards as expectShardNum.
-func (p *RandomBalancedShardPicker) PickShards(_ context.Context, snapshot metadata.Snapshot, expectShardNum int, enableDuplicateNode bool) ([]storage.ShardNode, error) {
+func (p *RandomBalancedShardPicker) PickShards(_ context.Context, snapshot metadata.Snapshot, expectShardNum int) ([]storage.ShardNode, error) {
 	shardNodes := snapshot.Topology.ClusterView.ShardNodes
 
 	nodeShardsMapping := make(map[string][]storage.ShardNode, 0)
 	for _, shardNode := range shardNodes {
 		_, exists := nodeShardsMapping[shardNode.NodeName]
 		if !exists {
-			shardNodes := []storage.ShardNode{}
-			nodeShardsMapping[shardNode.NodeName] = shardNodes
+			nodeShardsMapping[shardNode.NodeName] = []storage.ShardNode{}
 		}
 		nodeShardsMapping[shardNode.NodeName] = append(nodeShardsMapping[shardNode.NodeName], shardNode)
 	}
 
-	if !enableDuplicateNode {
-		if len(nodeShardsMapping) < expectShardNum {
-			return nil, errors.WithMessagef(ErrNodeNumberNotEnough, "number of nodes is:%d, expecet number of shards is:%d", len(nodeShardsMapping), expectShardNum)
-		}
-	}
-
 	// Try to make shards on different nodes.
 	result := make([]storage.ShardNode, 0, expectShardNum)
 	totalShardLength := len(shardNodes)
 	tempNodeShardMapping := make(map[string][]storage.ShardNode, len(nodeShardsMapping))
-	for {
-		nodeNames := make([]string, 0, len(nodeShardsMapping))
-		for nodeName := range nodeShardsMapping {
-			nodeNames = append(nodeNames, nodeName)
+	nodeNames := make([]string, 0, len(nodeShardsMapping))
+
+	for i := 0; i < expectShardNum; i++ {
+		// Initialize nodeNames.
+		if len(nodeNames) == 0 {
+			for nodeName := range nodeShardsMapping {
+				nodeNames = append(nodeNames, nodeName)
+			}
 		}
 
-		// Reset node shards when shard is all picked.
-		if len(result)%totalShardLength == 0 {
+		// Initialize tempNodeShardMapping in following cases:
+		// 1. First time to pick shards.
+		// 2. All shards are already picked.
+		if len(result) == 0 || len(result) == totalShardLength {
```

Comment:
```suggestion
		if len(result)%totalShardLength == 0
```
We cannot use this judgment condition to judge whether the current shard has been fetched, for example, there are 4 shards now, and the goal is to pick 12 shards. We need to reset when 8 shards are picked, but the writing method here is unsatisfactory.